controller/src/main/java/org/jboss/as/controller/CapabilityReferenceRecorder.java

```diff
@@ -321,4 +330,97 @@ protected String getRequirementName(OperationContext context, Resource resource,
+            this.dynamicRequirementNameMapper = dynamicRequirementNameMapper;
+            this.baseRequirementName = baseRequirementName;
+            this.baseDependentName = baseDependentName;
+        }
+
+        public ResourceCapabilityReferenceRecorder(Function<PathAddress, String[]> dynamicDependentNameMapper, String baseDependentName, String baseRequirementName) {
+            this(dynamicDependentNameMapper, baseDependentName, null, baseRequirementName);
+        }
+
+        public ResourceCapabilityReferenceRecorder(String baseDependentName, Function<PathAddress, String[]> dynamicRequirementNameMapper,String baseRequirementName) {
+            this(null, baseDependentName, dynamicRequirementNameMapper, baseRequirementName);
+        }
+
+        @Override
+        public void addCapabilityRequirements(OperationContext context, Resource resource, String attributeName, String... attributeValues) {
+            processCapabilityRequirement(context, attributeName, false);
+        }
+
+        @Override
+        public void removeCapabilityRequirements(OperationContext context, Resource resource, String attributeName, String... attributeValues) {
+            processCapabilityRequirement(context, attributeName, true);
+        }
+
+        private void processCapabilityRequirement(OperationContext context, String attributeName, boolean remove) {
+            String dependentName = getDependentName(context.getCurrentAddress());
+            String requirementName = getRequirementName(context.getCurrentAddress());
+            if (remove) {
+                context.deregisterCapabilityRequirement(requirementName, dependentName);
+            } else {
+                context.registerAdditionalCapabilityRequirement(requirementName, dependentName, attributeName);
+            }
+        }
+
+        private String getDependentName(PathAddress address) {
+            if (dynamicDependentNameMapper != null) {
+                return RuntimeCapability.buildDynamicCapabilityName(baseDependentName, dynamicDependentNameMapper.apply(address));
+            }
+            return baseDependentName;
+        }
+
+        private String getRequirementName(PathAddress address) {
+            if (dynamicRequirementNameMapper != null) {
+                return RuntimeCapability.buildDynamicCapabilityName(baseRequirementName, dynamicRequirementNameMapper.apply(address));
+            }
+            return baseRequirementName;
+        }
+
+        @Override
+        public String getBaseDependentName() {
+            return baseDependentName;
+        }
+
+        @Override
+        public String getBaseRequirementName() {
+            return baseRequirementName;
+        }
+
+        @Override
+        public boolean isDynamicDependent() {
+            return dynamicDependentNameMapper != null;
+        }
+
+        @Override
+        public String[] getRequirementPatternSegments(String dynamicElement, PathAddress registrationAddress) {
```

Comment:
We should be able to just return: dynamicRequirementNameMapper.apply(registrationAddress)
This method is always invoked with a null dynamicElement.
Additionally, dynamicRequirementNameMapper should never be null (see my comment above).

Reply:
We can't because the address might be aliased

Reply:
I'm a little confused.  If getRequirementPatternSegments(...) is called with a null dynamicElement (which is how it is called when building the resource description), this method just returns the value of dynamicRequirementNameMapper.apply(registrationAddress);
I don't see any special handling for aliased addresses here.

Reply:
I've updated the code in the meanwhile